DeepSpeech.ipynb

```diff
@@ -1320,170 +1292,390 @@
-    "        \n",
-    "        # Requirements to display a WER report\n",
-    "        if is_display_step:\n",
-    "            # Reset accuracy\n",
-    "            total_accuracy = 0.0\n",
-    "            # Create training results tuple\n",
-    "            train_results = ([],[],[],[])\n",
-    "            # Extend the session.run parameters\n",
-    "            params.append(train_results_params)\n",
-    "            params.append(avg_accuracy)\n",
-    "\n",
-    "        # Loop over the batches\n",
-    "        for batch in range(int(ceil(batches_per_device))):\n",
-    "            extra_params = { }\n",
-    "            if do_fulltrace:\n",
-    "                loss_run_metadata            = tf.RunMetadata()\n",
-    "                extra_params['options']      = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)\n",
-    "                extra_params['run_metadata'] = loss_run_metadata\n",
-    "            \n",
-    "            # Compute the batch\n",
-    "            training_time = stopwatch(training_time)\n",
-    "            result = session.run(params, feed_dict_train, **extra_params)\n",
-    "            training_time = stopwatch(training_time)\n",
-    "            \n",
-    "            # Add batch to loss\n",
-    "            loss += result[1]\n",
+    "        is_validation_step = (epoch > 0 and (epoch + 1) % validation_step == 0) or epoch == training_iters - 1\n",
     "\n",
-    "            # Log all variable states in current step\n",
-    "            step = epoch * total_batches + batch * len(available_devices)\n",
-    "            writer.add_summary(result[2], step)\n",
-    "            if do_fulltrace:\n",
-    "                writer.add_run_metadata(loss_run_metadata, 'loss_epoch%d_batch%d'   % (epoch, batch))\n",
-    "            writer.flush()\n",
-    "            \n",
-    "            # Collect individual sample results for WER report\n",
-    "            if is_display_step:\n",
-    "                collect_results(train_results, result[3])\n",
-    "                # Add batch to total_accuracy\n",
-    "                total_accuracy += result[4]\n",
-    "        \n",
-    "        # Print WER report\n",
-    "        if is_display_step:\n",
-    "            train_wer = calculate_and_print_wer_report( \\\n",
-    "                \"Training avg_cer=\" + \"{:.9f}\".format((total_accuracy / ceil(batches_per_device))), train_results)\n",
-    "            print\n",
+    "        print \"Training model...\"\n",
+    "        train_time = stopwatch(train_time)\n",
+    "        result = apply_batches(train_context, session, epoch=epoch, query_report=is_display_step)\n",
+    "        train_time = stopwatch(train_time)\n",
+    "\n",
+    "        result = print_report(\"Training\", result)\n",
+    "        # If there was a WER calculated, we keep it\n",
+    "        train_wer = result if result is not None else train_wer\n",
```

Comment:
As `is_display_step` need not be true for the final epoch, `train_wer` will not reflect the final WER.